lumigator/python/mzai/backend/backend/services/jobs.py

```diff
@@ -311,10 +320,15 @@ def create_job(
         entrypoint = RayJobEntrypoint(
             config=ray_config, metadata=metadata, runtime_env=runtime_env, num_gpus=worker_gpus
         )
-        loguru.logger.info("Submitting Ray job...")
+        loguru.logger.info("Submitting {job_type} Ray job...")
         submit_ray_job(self.ray_client, entrypoint)
 
-        background_tasks.add_task(self._watch_job, record.id, request)
+        if job_type == JobType.INFERENCE:
+            # Inference jobs produce a new dataset
+            # Add the dataset to the (local) database
+            background_tasks.add_task(
+                self._run_after_job, record.id, self._add_dataset_to_db, record.id, request
```

Comment:
This is the only bit I'm not too comfortable with. We now have a job modifying a queue created at experiment (i.e. orchestrator) level.

On the one side, I understand it can only add jobs, never amend or remove, and it removes the need for the orchestrator to know details (e.g. output data fields) of the job.

On the other, which I feel is stronger, it smears the responsibility division between orchestrator and job, forcing the job creator to know how lumigator task queues and, say, data storage in the DB work. If/when anything in the flow fails, it would also be a bit harder to debug...

Seeing the need to start testing this workflow/composite task and the fact we're discussing full orchestration elsewhere, perhaps what belongs in this PR is the outline of what problem keeping all modifications of the background_tasks at experiment level has.

WDYT?

Reply:
I see your point and I agree we can start using it at experiment level, then verify if and how much it limits us. I have added a few pros/cons to the related commit to summarize my experience with it